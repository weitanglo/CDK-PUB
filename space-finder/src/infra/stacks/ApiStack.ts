import { Stack, StackProps } from 'aws-cdk-lib'
import { AuthorizationType, CognitoUserPoolsAuthorizer, Cors, LambdaIntegration, MethodLoggingLevel, MethodOptions, ResourceOptions, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { IUserPool } from 'aws-cdk-lib/aws-cognito';
import { Construct } from 'constructs';

interface ApiStackProps extends StackProps {
    spacesLambdaIntegration: LambdaIntegration,
    userPool: IUserPool;
}

export class ApiStack extends Stack {

    constructor(scope: Construct, id: string, props: ApiStackProps) {
        super(scope, id, props);

        const api = new RestApi(this, 'SpacesApi');

        const authorizer = new CognitoUserPoolsAuthorizer(this, 'SpacesApiAuthorizer', {
            cognitoUserPools: [props.userPool],
            identitySource: 'method.request.header.Authorization'    //we will do it differently from default AWS, this is the location of the authorization header.
        });
        authorizer._attachToApi(api)                       //authorizer cant exist without an APi, without it call synth will have a error

        const optionsWithAuth: MethodOptions = {
            authorizationType: AuthorizationType.COGNITO,
            authorizer: {
                authorizerId: authorizer.authorizerId
            }
        }

        const optionsWithCros: ResourceOptions = {
            defaultCorsPreflightOptions: {
                allowOrigins: Cors.ALL_ORIGINS,
                allowMethods: Cors.ALL_METHODS
            }
        }

        const spacesResource = api.root.addResource('spaces', optionsWithCros);
        spacesResource.addMethod('GET', props.spacesLambdaIntegration, optionsWithAuth);
        spacesResource.addMethod('POST', props.spacesLambdaIntegration, optionsWithAuth);
        spacesResource.addMethod('PUT', props.spacesLambdaIntegration, optionsWithAuth);
        spacesResource.addMethod('DELETE', props.spacesLambdaIntegration, optionsWithAuth);
    };
}
