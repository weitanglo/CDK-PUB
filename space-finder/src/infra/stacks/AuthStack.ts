import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib'
import { Construct } from 'constructs';
import { CfnIdentityPool, CfnIdentityPoolRoleAttachment, CfnUserPoolGroup, UserPool, UserPoolClient } from 'aws-cdk-lib/aws-cognito';
import { Effect, FederatedPrincipal, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { IBucket } from 'aws-cdk-lib/aws-s3';



interface AuthStackProps extends StackProps {
    photosBucket: IBucket
}


export class AuthStack extends Stack {

    public userPool: UserPool;
    private userPoolClient: UserPoolClient;
    private identityPool: CfnIdentityPool;
    private authenticatedRole: Role;
    private unAuthenticatedRole: Role;
    private adminRole: Role;

    constructor(scope: Construct, id: string, props: AuthStackProps) {
        super(scope, id, props);

        this.createUserPool();
        this.createUserPoolClient();
        this.createIdentityPool();
        this.createRoles(props.photosBucket);
        this.attachRoles();
        this.createAdminsGroup();   //After create Role then create group, if not the role will not connect to group
    }

    private createUserPool() {
        this.userPool = new UserPool(this, 'SpaceUserPool', {
            selfSignUpEnabled: true,
            signInAliases: {
                username: true,
                email: true
            }
        });
        new CfnOutput(this, 'SpaceUserPoolId', {
            value: this.userPool.userPoolId
        })
    }
    private createUserPoolClient() {
        this.userPoolClient = this.userPool.addClient('SpaceUserPoolClient', {
            authFlows: {
                adminUserPassword: true,
                custom: true,
                userPassword: true,
                userSrp: true
            }
        });
        new CfnOutput(this, 'SpaceUserPoolClientId', {
            value: this.userPoolClient.userPoolClientId
        })
    }

    private createAdminsGroup() {
        new CfnUserPoolGroup(this, 'SpaceAdmins', {
            userPoolId: this.userPool.userPoolId,
            groupName: 'admin',
            roleArn: this.adminRole.roleArn          //connect adminRole to this group 
        })
    }

    private createIdentityPool() {
        this.identityPool = new CfnIdentityPool(this, 'SpaceIdentityPool', {
            allowUnauthenticatedIdentities: true,
            cognitoIdentityProviders: [{
                clientId: this.userPoolClient.userPoolClientId,
                providerName: this.userPool.userPoolProviderName
            }]
        })
        new CfnOutput(this, 'SpaceIdentityPoolRef', {
            value: this.identityPool.ref             //because using Cfn, we can get ref to access identity pool ID
        })
    }

    private createRoles(photosBucket: IBucket) {
        this.authenticatedRole = new Role(this, 'CognitoDefaultAuthenticatedRole', {
            assumedBy: new FederatedPrincipal('cognito-identity.amazonaws.com', {
                StringEquals: {                                               
                    'cognito-identity.amazonaws.com:aud': this.identityPool.ref
                },
                'ForAnyValue:StringLike': {
                    'cognito-identity.amazonaws.com:amr': 'authenticated'
                }
            },
                'sts:AssumeRoleWithWebIdentity'
            )
        });
        this.unAuthenticatedRole = new Role(this, 'CognitoDefaultUnauthenticatedRole', {
            assumedBy: new FederatedPrincipal('cognito-identity.amazonaws.com', {
                StringEquals: {
                    'cognito-identity.amazonaws.com:aud': this.identityPool.ref
                },
                'ForAnyValue:StringLike': {
                    'cognito-identity.amazonaws.com:amr': 'unauthenticated'
                }
            },
                'sts:AssumeRoleWithWebIdentity'
            )
        });
        this.adminRole = new Role(this, 'CognitoAdminRole', {
            assumedBy: new FederatedPrincipal('cognito-identity.amazonaws.com', {
                StringEquals: {
                    'cognito-identity.amazonaws.com:aud': this.identityPool.ref
                },
                'ForAnyValue:StringLike': {
                    'cognito-identity.amazonaws.com:amr': 'authenticated'
                }
            },
                'sts:AssumeRoleWithWebIdentity'
            )
        });
        this.adminRole.addToPolicy(new PolicyStatement({
            effect:Effect.ALLOW,
            actions:[
                's3:ListAllMyBuckets',
                's3:PutObject',
                's3:PutObjectAcl'
            ],
            resources:[photosBucket.bucketArn + '/*']
        }))
    }

    private attachRoles(){
        new CfnIdentityPoolRoleAttachment(this, 'RolesAttachment', {
            identityPoolId: this.identityPool.ref,
            roles:{
                'authenticated': this.authenticatedRole.roleArn,
                'unauthenticated': this.unAuthenticatedRole.roleArn
            },
            roleMappings:{
                adminsMapping: {
                    type: 'Token',
                    ambiguousRoleResolution: 'AuthenticatedRole',  //mean that two roles
                    identityProvider: `${this.userPool.userPoolProviderName}:${this.userPoolClient.userPoolClientId}`
                }
            }
        })
    }

}
