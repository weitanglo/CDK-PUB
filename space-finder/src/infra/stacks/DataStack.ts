import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib'
import { AttributeType, ITable, Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { getSuffixFromStack } from '../Utils';
import { Bucket, HttpMethods, IBucket, ObjectOwnership } from 'aws-cdk-lib/aws-s3';

export class DataStack extends Stack {

    public readonly spacesTable: ITable
    public readonly photosBucket: IBucket

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const suffix = getSuffixFromStack(this);

        this.photosBucket = new Bucket(this, 'spaceFinderPhotos', {
            bucketName: `space-finder-photo-${suffix}`,
            cors: [{
                allowedMethods: [
                    HttpMethods.HEAD,
                    HttpMethods.GET,
                    HttpMethods.PUT
                ],
                allowedOrigins: ['*'],
                allowedHeaders: ['*'],
            }],
            // accessControl: BucketAccessControl.PUBLIC_READ,  //DOC recommanded but currently not working
            objectOwnership: ObjectOwnership.OBJECT_WRITER,
            blockPublicAccess: {
                blockPublicAcls: false,
                blockPublicPolicy: false,
                ignorePublicAcls: false,
                restrictPublicBuckets: false,
            }
        });

        new CfnOutput(this, 'SpaceFinderPhotosBucketName', {
            value: this.photosBucket.bucketName
        });

        this.spacesTable = new Table(this, 'SpacesTable', {
            partitionKey: {
                name: 'id',
                type: AttributeType.STRING
            },
            tableName: `SpaceTable-${suffix}`
        });

        this.photosBucket = new Bucket(this, 'spaceFinderPhotos', {
            bucketName: `space-finder-photo-${suffix}`,
            cors: [{
                allowedMethods: [
                    HttpMethods.HEAD,
                    HttpMethods.GET,
                    HttpMethods.PUT
                ],
                allowedOrigins: ['*'],
                allowedHeaders: ['*']
            }],   
            // accessControl: BucketAccessControl.PUBLIC_READ,  //DOC recommanded but currently not working
            objectOwnership: ObjectOwnership.OBJECT_WRITER,
            blockPublicAccess: {
                blockPublicAcls: false,
                blockPublicPolicy: false,
                ignorePublicAcls: false,
                restrictPublicBuckets: false,
            }
        });
        
        new CfnOutput(this, 'SpaceFinderPhotosBucketName', {
            value: this.photosBucket.bucketName
        })
    };
}
