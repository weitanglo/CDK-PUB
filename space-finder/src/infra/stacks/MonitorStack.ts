import { Duration, Stack, StackProps } from "aws-cdk-lib";
import { Alarm, Metric, Unit } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Topic } from "aws-cdk-lib/aws-sns";
import { LambdaSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import { Construct } from "constructs";
import { join } from 'path';

export class MonitorStack extends Stack {

    constructor(scop: Construct, id: string, props?: StackProps) {
        super(scop, id, props);

        const webHookLambda = new NodejsFunction(this, 'webHookLambda', {
            runtime: Runtime.NODEJS_18_X,
            handler: 'handler',
            entry:(join(__dirname, '..', '..', 'services', 'monitor', 'handler.ts'))
        });

        const alarmTopic = new Topic(this, 'alarmTopic', {
            displayName: 'alarmTopic',
            topicName: 'alarmTopic'
        });
        alarmTopic.addSubscription(new LambdaSubscription(webHookLambda));

        const spacesApi4xxAlarm = new Alarm(this, 'spacesApi4xxAlarm', {
            metric: new Metric({
                metricName: '4XXError',
                namespace: 'AWS/ApiGateway',
                period: Duration.minutes(1),
                statistic: 'Sum',
                unit: Unit.COUNT,
                dimensionsMap: {
                    "ApiName": "SpacesApi"
                }
            }),
            evaluationPeriods: 1,
            threshold: 5,
            alarmName: 'spacesApi4xxAlarm'
        });
        const topicAction = new SnsAction(alarmTopic);
        spacesApi4xxAlarm.addAlarmAction(topicAction);
        spacesApi4xxAlarm.addOkAction(topicAction);
    }
}

