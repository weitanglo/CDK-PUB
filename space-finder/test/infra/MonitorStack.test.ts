import { App } from "aws-cdk-lib";
import { MonitorStack } from "../../src/infra/stacks/MonitorStack";
import { Capture, Match, Template } from "aws-cdk-lib/assertions";


describe('initial tes suite', () => {

    // arrange step ,when preparing our step environment
    let monitorStackTemplate: Template;

    beforeAll(() => {
        const testApp = new App({
            outdir: 'cdk.out'
        });
        const monitorStack = new MonitorStack(testApp, 'MonitorStack');
        monitorStackTemplate = Template.fromStack(monitorStack);
    })

    test('Lambda properties', () => {
        // act step ,when acting with our classes or our system under test
        // assert step ,when making sure that thing are the right way or they are not
        monitorStackTemplate.hasResourceProperties('AWS::Lambda::Function', {
            Handler: 'index.handler',
            Runtime: 'nodejs18.x'
        });
        // expect(true).toBeTruthy();
    });

    test('Sns topic properties', () => {
        monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
            DisplayName: 'alarmTopic',
            TopicName: 'alarmTopic'
        });
    });
    test('Sns Subscription properties - with CDK matchers', () => {
        monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription',
            Match.objectEquals(
                {
                    Protocol: 'lambda',
                    TopicArn: {
                        Ref: Match.stringLikeRegexp('alarmTopic')
                    },
                    Endpoint: {
                        'Fn::GetAtt': [
                            Match.stringLikeRegexp('webHookLambda'),
                            'Arn'
                        ]
                    }
                }));
    });

    test('Sns Subscription properties - with exact values', () => {
        const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
        console.log(snsTopic);
        const snsTopicName = Object.keys(snsTopic)[0];
        const lambda = monitorStackTemplate.findResources('AWS::Lambda::Function');
        const lambdaName = Object.keys(lambda)[0];

        monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription',
            Match.objectEquals(
                {
                    Protocol: 'lambda',
                    TopicArn: {
                        Ref: snsTopicName
                    },
                    Endpoint: {
                        'Fn::GetAtt': [
                            lambdaName,
                            'Arn'
                        ]
                    }
                }));
    });

    test('Alarm actions - with Capture and Jest matchers', ()=> {
        const alarmActionsCapture = new Capture();
        monitorStackTemplate.hasResourceProperties('AWS::CloudWatch::Alarm', {
            AlarmActions: alarmActionsCapture
        })

        expect(alarmActionsCapture.asArray()).toEqual([{
            Ref: expect.stringMatching(/^alarmTopic/)
        }])
    });

    test('Monitor stack snapshot', () => {
        expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
    });

    test('Lambda stack snapshot', () => {
        const lambda = monitorStackTemplate.findResources('AWS::Lambda::Function');
        expect(lambda).toMatchSnapshot();
    });

    test('Snstopic stack snapshot', () => {
        const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
        expect(snsTopic).toMatchSnapshot();
    });
})