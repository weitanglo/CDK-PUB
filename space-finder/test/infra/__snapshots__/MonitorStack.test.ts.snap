// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial tes suite Lambda stack snapshot 1`] = `
{
  "webHookLambda3427ABA6": {
    "DependsOn": [
      "webHookLambdaServiceRole1A032560",
    ],
    "Properties": {
      "Code": {
        "S3Bucket": {
          "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
        },
        "S3Key": "dee3b54a9a812d75cbae834cfbf1ba8b9c50401cdaf74f1b8f48dad7637be23a.zip",
      },
      "Environment": {
        "Variables": {
          "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
        },
      },
      "Handler": "index.handler",
      "Role": {
        "Fn::GetAtt": [
          "webHookLambdaServiceRole1A032560",
          "Arn",
        ],
      },
      "Runtime": "nodejs18.x",
    },
    "Type": "AWS::Lambda::Function",
  },
}
`;

exports[`initial tes suite Monitor stack snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "alarmTopic0C9A2A41": {
      "Properties": {
        "DisplayName": "alarmTopic",
        "TopicName": "alarmTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "spacesApi4xxAlarmA15E009C": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "alarmTopic0C9A2A41",
          },
        ],
        "AlarmName": "spacesApi4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpacesApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "OKActions": [
          {
            "Ref": "alarmTopic0C9A2A41",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 5,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "webHookLambda3427ABA6": {
      "DependsOn": [
        "webHookLambdaServiceRole1A032560",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "dee3b54a9a812d75cbae834cfbf1ba8b9c50401cdaf74f1b8f48dad7637be23a.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "webHookLambdaServiceRole1A032560",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "webHookLambdaAllowInvokeMonitorStackalarmTopic45B6AF8430DB74B1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "webHookLambda3427ABA6",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "alarmTopic0C9A2A41",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "webHookLambdaServiceRole1A032560": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "webHookLambdaalarmTopic761A88CB": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "webHookLambda3427ABA6",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "alarmTopic0C9A2A41",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`initial tes suite Snstopic stack snapshot 1`] = `
{
  "alarmTopic0C9A2A41": {
    "Properties": {
      "DisplayName": "alarmTopic",
      "TopicName": "alarmTopic",
    },
    "Type": "AWS::SNS::Topic",
  },
}
`;
